File: src/lever_mcp/gmail_client.py
"""
Gmail client with OAuth 2.0 support.
Handles authentication and email sending via Gmail API.
"""
import logging
import base64
from typing import Optional, Dict, Any
from google.oauth2.credentials import Credentials
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from lever_mcp.oauth_config import oauth_config, GMAIL_SCOPES

logger = logging.getLogger(__name__)


class GmailClient:
    """Client for interacting with Gmail API using OAuth 2.0."""
    
    def __init__(self, access_token: Optional[str] = None, user_id: str = "default"):
        """
        Initialize Gmail client.
        
        Args:
            access_token: OAuth access token from the agent (on-behalf-of flow)
            user_id: User identifier for token storage
        """
        self.user_id = user_id
        self.credentials = None
        
        if access_token:
            # Use token provided by agent (on-behalf-of flow)
            self.credentials = Credentials(
                token=access_token,
                scopes=GMAIL_SCOPES
            )
        else:
            # Try to load stored token
            self._load_credentials()
    
    def _load_credentials(self) -> None:
        """Load credentials from storage."""
        token_data = oauth_config.load_token(self.user_id)
        
        if token_data:
            self.credentials = Credentials.from_authorized_user_info(
                token_data,
                GMAIL_SCOPES
            )
            
            # Refresh token if expired
            if self.credentials and self.credentials.expired and self.credentials.refresh_token:
                try:
                    self.credentials.refresh(Request())
                    # Save refreshed token
                    self._save_credentials()
                    logger.info("Token refreshed successfully")
                except Exception as e:
                    logger.error(f"Failed to refresh token: {e}")
                    self.credentials = None
    
    def _save_credentials(self) -> None:
        """Save credentials to storage."""
        if self.credentials:
            token_data = {
                'token': self.credentials.token,
                'refresh_token': self.credentials.refresh_token,
                'token_uri': self.credentials.token_uri,
                'client_id': self.credentials.client_id,
                'client_secret': self.credentials.client_secret,
                'scopes': self.credentials.scopes
            }
            oauth_config.save_token(token_data, self.user_id)
    
    def is_authenticated(self) -> bool:
        """Check if client has valid credentials."""
        return self.credentials is not None and self.credentials.valid
    
    def set_token(self, token_data: Dict[str, Any]) -> None:
        """
        Set OAuth token from agent.
        
        Args:
            token_data: Token data containing access_token and optionally refresh_token
        """
        if isinstance(token_data, str):
            # Simple access token string
            self.credentials = Credentials(
                token=token_data,
                scopes=GMAIL_SCOPES
            )
        elif isinstance(token_data, dict):
            # Full token data
            if 'access_token' in token_data:
                self.credentials = Credentials(
                    token=token_data['access_token'],
                    refresh_token=token_data.get('refresh_token'),
                    token_uri=token_data.get('token_uri', 'https://oauth2.googleapis.com/token'),
                    client_id=token_data.get('client_id', oauth_config.client_id),
                    client_secret=token_data.get('client_secret', oauth_config.client_secret),
                    scopes=GMAIL_SCOPES
                )
                # Save for future use
                self._save_credentials()
            else:
                raise ValueError("Token data must contain 'access_token'")
        else:
            raise ValueError("Token must be a string or dictionary")
    
    async def send_email(
        self,
        to: str,
        subject: str,
        body: str,
        cc: Optional[str] = None,
        bcc: Optional[str] = None,
        is_html: bool = True
    ) -> Dict[str, Any]:
        """
        Send an email via Gmail API.
        
        Args:
            to: Recipient email address
            subject: Email subject
            body: Email body (HTML or plain text)
            cc: CC recipients (comma-separated)
            bcc: BCC recipients (comma-separated)
            is_html: Whether body is HTML
            
        Returns:
            Response from Gmail API
            
        Raises:
            ValueError: If not authenticated
            HttpError: If Gmail API request fails
        """
        if not self.is_authenticated():
            raise ValueError(
                "Not authenticated. Please provide an OAuth token or complete authentication flow."
            )
        
        try:
            # Build Gmail service
            service = build('gmail', 'v1', credentials=self.credentials)
            
            # Create message
            message_parts = [
                f"To: {to}",
                f"Subject: {subject}",
                "MIME-Version: 1.0",
            ]
            
            if is_html:
                message_parts.append("Content-Type: text/html; charset=utf-8")
            else:
                message_parts.append("Content-Type: text/plain; charset=utf-8")
            
            if cc:
                message_parts.append(f"Cc: {cc}")
            if bcc:
                message_parts.append(f"Bcc: {bcc}")
            
            message_parts.append("")  # Empty line between headers and body
            message_parts.append(body)
            
            raw_message = "\n".join(message_parts)
            
            # Encode message
            encoded_message = base64.urlsafe_b64encode(
                raw_message.encode('utf-8')
            ).decode('utf-8')
            
            # Send message
            message = service.users().messages().send(
                userId='me',
                body={'raw': encoded_message}
            ).execute()
            
            logger.info(f"Email sent successfully. Message ID: {message['id']}")
            
            return {
                "status": "success",
                "message_id": message['id'],
                "to": to,
                "subject": subject
            }
            
        except HttpError as error:
            logger.error(f"Gmail API error: {error}")
            raise
        except Exception as error:
            logger.error(f"Error sending email: {error}")
            raise
    
    def get_auth_url(self) -> str:
        """
        Get OAuth authorization URL for user to authenticate.
        
        Returns:
            Authorization URL
        """
        from google_auth_oauthlib.flow import Flow
        
        if not oauth_config.is_configured():
            raise ValueError("OAuth not configured")
        
        flow = Flow.from_client_config(
            oauth_config.get_client_config(),
            scopes=GMAIL_SCOPES,
            redirect_uri=oauth_config.redirect_uri
        )
        
        auth_url, _ = flow.authorization_url(
            access_type='offline',
            include_granted_scopes='true',
            prompt='consent'
        )
        
        return auth_url
    
    def exchange_code_for_token(self, code: str) -> Dict[str, Any]:
        """
        Exchange authorization code for access token.
        
        Args:
            code: Authorization code from OAuth callback
            
        Returns:
            Token data
        """
        from google_auth_oauthlib.flow import Flow
        
        if not oauth_config.is_configured():
            raise ValueError("OAuth not configured")
        
        flow = Flow.from_client_config(
            oauth_config.get_client_config(),
            scopes=GMAIL_SCOPES,
            redirect_uri=oauth_config.redirect_uri
        )
        
        flow.fetch_token(code=code)
        
        self.credentials = flow.credentials
        self._save_credentials()
        
        return {
            'access_token': self.credentials.token,
            'refresh_token': self.credentials.refresh_token,
            'expires_in': self.credentials.expiry.timestamp() if self.credentials.expiry else None
        }

File: src/mcp/ts/src/gmail-client.ts
/**
 * Gmail client with OAuth 2.0 support.
 * Handles authentication and email sending via Gmail API.
 */
import { google } from 'googleapis';
import { OAuth2Client } from 'google-auth-library';
import { oauthConfig, GMAIL_SCOPES, TokenData } from './oauth-config.js';

export interface EmailOptions {
    to: string;
    subject: string;
    body: string;
    cc?: string;
    bcc?: string;
    isHtml?: boolean;
}

export interface EmailResult {
    status: string;
    messageId: string;
    to: string;
    subject: string;
}

export class GmailClient {
    private userId: string;
    private oauth2Client: OAuth2Client | null = null;

    constructor(accessToken?: string, userId: string = 'default') {
        this.userId = userId;

        if (accessToken) {
            // Use token provided by agent (on-behalf-of flow)
            this.oauth2Client = new OAuth2Client(
                oauthConfig.clientId,
                oauthConfig.clientSecret,
                oauthConfig.redirectUri
            );
            this.oauth2Client.setCredentials({
                access_token: accessToken,
                scope: GMAIL_SCOPES.join(' ')
            });
        } else {
            // Try to load stored token
            this.loadCredentials();
        }
    }

    private async loadCredentials(): Promise<void> {
        const tokenData = await oauthConfig.loadToken(this.userId);

        if (tokenData) {
            this.oauth2Client = new OAuth2Client(
                oauthConfig.clientId,
                oauthConfig.clientSecret,
                oauthConfig.redirectUri
            );

            this.oauth2Client.setCredentials({
                access_token: tokenData.access_token || tokenData.token,
                refresh_token: tokenData.refresh_token,
                scope: GMAIL_SCOPES.join(' ')
            });

            // Refresh token if needed
            try {
                const { credentials } = await this.oauth2Client.refreshAccessToken();
                this.oauth2Client.setCredentials(credentials);
                
                // Save refreshed token
                await this.saveCredentials(credentials);
                console.log('Token refreshed successfully');
            } catch (error) {
                console.error('Failed to refresh token:', error);
                this.oauth2Client = null;
            }
        }
    }

    private async saveCredentials(credentials: any): Promise<void> {
        if (this.oauth2Client && credentials) {
            const tokenData: TokenData = {
                access_token: credentials.access_token,
                refresh_token: credentials.refresh_token,
                token_uri: credentials.token_uri,
                client_id: oauthConfig.clientId,
                client_secret: oauthConfig.clientSecret,
                scopes: GMAIL_SCOPES
            };
            await oauthConfig.saveToken(tokenData, this.userId);
        }
    }

    isAuthenticated(): boolean {
        return this.oauth2Client !== null && Boolean(this.oauth2Client.credentials.access_token);
    }

    async setToken(tokenData: TokenData | string): Promise<void> {
        if (typeof tokenData === 'string') {
            // Simple access token string
            this.oauth2Client = new OAuth2Client(
                oauthConfig.clientId,
                oauthConfig.clientSecret,
                oauthConfig.redirectUri
            );
            this.oauth2Client.setCredentials({
                access_token: tokenData,
                scope: GMAIL_SCOPES.join(' ')
            });
        } else if (typeof tokenData === 'object' && tokenData !== null) {
            // Full token data
            if (tokenData.access_token) {
                this.oauth2Client = new OAuth2Client(
                    tokenData.client_id || oauthConfig.clientId,
                    tokenData.client_secret || oauthConfig.clientSecret,
                    oauthConfig.redirectUri
                );
                this.oauth2Client.setCredentials({
                    access_token: tokenData.access_token,
                    refresh_token: tokenData.refresh_token,
                    scope: GMAIL_SCOPES.join(' ')
                });
                // Save for future use
                await this.saveCredentials(this.oauth2Client.credentials);
            } else {
                throw new Error("Token data must contain 'access_token'");
            }
        } else {
            throw new Error('Token must be a string or object');
        }
    }

    async sendEmail(options: EmailOptions): Promise<EmailResult> {
        if (!this.isAuthenticated()) {
            throw new Error(
                'Not authenticated. Please provide an OAuth token or complete authentication flow.'
            );
        }

        try {
            // Build Gmail service
            const gmail = google.gmail({ version: 'v1', auth: this.oauth2Client! });

            // Create message
            const messageParts = [
                `To: ${options.to}`,
                `Subject: ${options.subject}`,
                'MIME-Version: 1.0',
            ];

            if (options.isHtml !== false) {
                messageParts.push('Content-Type: text/html; charset=utf-8');
            } else {
                messageParts.push('Content-Type: text/plain; charset=utf-8');
            }

            if (options.cc) {
                messageParts.push(`Cc: ${options.cc}`);
            }
            if (options.bcc) {
                messageParts.push(`Bcc: ${options.bcc}`);
            }

            messageParts.push(''); // Empty line between headers and body
            messageParts.push(options.body);

            const rawMessage = messageParts.join('\n');

            // Encode message
            const encodedMessage = Buffer.from(rawMessage, 'utf-8')
                .toString('base64')
                .replace(/\+/g, '-')
                .replace(/\//g, '_')
                .replace(/=+$/, '');

            // Send message
            const response = await gmail.users.messages.send({
                userId: 'me',
                requestBody: {
                    raw: encodedMessage
                }
            });

            console.log(`Email sent successfully. Message ID: ${response.data.id}`);

            return {
                status: 'success',
                messageId: response.data.id || '',
                to: options.to,
                subject: options.subject
            };

        } catch (error: any) {
            console.error('Gmail API error:', error);
            throw error;
        }
    }

    getAuthUrl(): string {
        if (!oauthConfig.isConfigured()) {
            throw new Error('OAuth not configured');
        }

        const oauth2Client = new OAuth2Client(
            oauthConfig.clientId,
            oauthConfig.clientSecret,
            oauthConfig.redirectUri
        );

        const authUrl = oauth2Client.generateAuthUrl({
            access_type: 'offline',
            scope: GMAIL_SCOPES,
            include_granted_scopes: true,
            prompt: 'consent'
        });

        return authUrl;
    }

    async exchangeCodeForToken(code: string): Promise<TokenData> {
        if (!oauthConfig.isConfigured()) {
            throw new Error('OAuth not configured');
        }

        const oauth2Client = new OAuth2Client(
            oauthConfig.clientId,
            oauthConfig.clientSecret,
            oauthConfig.redirectUri
        );

        const { tokens } = await oauth2Client.getToken(code);
        
        this.oauth2Client = oauth2Client;
        this.oauth2Client.setCredentials(tokens);
        await this.saveCredentials(tokens);

        return {
            access_token: tokens.access_token || undefined,
            refresh_token: tokens.refresh_token || undefined,
            expires_in: tokens.expiry_date ? Math.floor((tokens.expiry_date - Date.now()) / 1000) : undefined,
            token_type: tokens.token_type || 'Bearer',
            scope: tokens.scope
        };
    }
}
